### PR TITLE
WIP - Update to Documentation - Bert

### PR DESCRIPTION
**Pull request by koaning**
**Proposed changes**:
- Add a new part to the documentation page that explains how to use Bert. 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

@dakshvar22 This is what it looks like on my side. 

image

### CODE REVIEW

**Comment by koaning** [comment]
But somehow that link is not showing up when you're on another page. Any idea where to set that? 

image

File: docs/nlu/using-bert.rst [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    # output from the heavy model
+    2020-03-30 16:47:04 INFO     rasa.nlu.model  - Starting to train component DIETClassifier
+    Epochs: 100%|███████████████████████████████| 50/50 [04:33<00:00,  ...]
+    2020-03-30 16:49:52 INFO     rasa.nlu.test  - Running model for predictions:
+    100%|███████████████████████████████████████| 2396/2396 [07:20<00:00,  5.69it/s]
+
+From the logs we can gather an important observation. 
+The heavy model is a fair bit slower, not in training, but at inference time
+we see a ~6 fold increase. Depending on your use-case this is 
+something to seriously consider.
```
Comment (dakshvar22) [inline docs/nlu/using-bert.rst]: This is an important observation. Can we highlight this in some manner. For some reason, I just skipped this observation in my first reading. Two options - 
1. Move it to the observations section below, so that all observations are in one place and reader doesn't miss them
2. Or create a `note` block and put this in that so that it is highlighted.
What do you think?

File: docs/nlu/using-bert.rst [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+On all fronts we see that the model with the Bert embeddings performs better. 
+But it deserves mentioning that the effect is more pronounced in the entities.
+Note that these results may not be the same on your use-case. Every assistant 
+is different so it is important that you keep comparing. 
+
+It also deserves 
+mentioning that you need to beware that you don't over-optimise training data
+that you've generated yourself. End users will use the assistant in ways you 
+probably did not anticipate. Typically it is more important to gather data of 
+actual users than it is to get the best F1 score on an artificial dataset.
```
Comment (dakshvar22) [inline docs/nlu/using-bert.rst]: Is this related to this doc page? I didn't understand what you mean here as well.

File: docs/nlu/using-bert.rst [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+.. image:: /_static/images/bert-entities.png
+
+Observations 
+~~~~~~~~~~~~
+
+On all fronts we see that the model with the Bert embeddings performs better. 
+But it deserves mentioning that the effect is more pronounced in the entities.
+Note that these results may not be the same on your use-case. Every assistant 
+is different so it is important that you keep comparing.
```
Comment (dakshvar22) [inline docs/nlu/using-bert.rst]:
```suggestion
On all fronts we see that the pipeline with pre-trained embeddings from ``BERT`` performs better than the lightweight configuration consisting of ``CountVectorsFeaturizer``.
Specifically, the performance gain for intent classification is marginal, although for entity recognition it is more pronounced. Note that these results may not be the same on your use-case. Every assistant is different so it is important that you keep comparing. 
```

Comment (dakshvar22) [inline docs/nlu/using-bert.rst]: Can we motivate this point a bit better, making it explicit that there are two trade-offs at play here - 
1. Which task is more important - intent classification or entity recognition?
2. Is accuracy more important or latency of the bot during prediction.

File: docs/nlu/using-bert.rst [hunk 4]
```diff
@@ -0,0 +1,156 @@
+:desc: Find out how to use only Bert inside of Rasa NLU.
+
+Using Bert
```
[the remark this answers is not in the document]
Comment (koaning) [reply docs/nlu/using-bert.rst]: I'm not a SEO expert. My guideline is that this page needs to serve the purpose of clear explainin'. Feels approrpiate to primarily focus on that.

File: docs/nlu/using-bert.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    language: en
+    pipeline:
+    - name: WhitespaceTokenizer
+    - name: CountVectorsFeaturizer
+    - name: CountVectorsFeaturizer
+    analyzer: char_wb
+    min_ngram: 1
+    max_ngram: 4
+    - name: DIETClassifier
+    epochs: 20
```
[the remark this answers is not in the document]
Comment (koaning) [reply docs/nlu/using-bert.rst]: well spotted.

File: docs/nlu/using-bert.rst (hunk 2 above)
Comment (koaning) [reply docs/nlu/using-bert.rst]: rephrased it.

File: docs/nlu/using-bert.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (amn41) [reply docs/nlu/using-bert.rst]: +1 to Daksh's comment!

File: docs/nlu/using-bert.rst [hunk 6]
```diff
@@ -0,0 +1,157 @@
+:desc: Find out how to use only Bert inside of Rasa NLU.
```
Comment (amn41) [inline docs/nlu/using-bert.rst]: this goes into the meta tag and is important for SEO. Can be a few sentences, and should also include other important keywords like GPT-2. @DominikRos has experience writing these

File: docs/nlu/using-bert.rst (hunk 1 above)
Comment (koaning) [reply docs/nlu/using-bert.rst]: I've added a note and moved it just before the results.

File: docs/nlu/using-bert.rst (hunk 6 above)
Comment (koaning) [reply docs/nlu/using-bert.rst]: I've added Huggingface now too. Would love to hear about more buzzwords to add.

File: docs/nlu/using-bert.rst (hunk 3 above)
Comment (koaning) [reply docs/nlu/using-bert.rst]: I've expanded the points a bit.

File: docs/nlu/using-bert.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (dakshvar22) [reply docs/nlu/using-bert.rst]: Changed it to "Using BERT inside Rasa NLU"

File: docs/nlu/using-bert.rst [hunk 7]
```diff
@@ -0,0 +1,185 @@
+:desc: Find out how to use pre-trained embeddings from language models implemented in HuggingFace's Transformers
```
Comment (amn41) [inline docs/nlu/using-bert.rst]: let's throw 'chatbot' in here

File: docs/nlu/using-bert.rst [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+       like BERT, GPT-2, etc. inside Rasa NLU.
+
+Using BERT inside Rasa NLU
+==========================
+
+Since Rasa 1.8.0, you can use pre-trained embeddings from language models like
+`BERT <https://arxiv.org/abs/1810.04805>`_ inside of Rasa NLU pipelines.
+The goal of this document is to show how you can do that
+as well as some practical tips on exploring these new tools.
+
```
Comment (amn41) [inline docs/nlu/using-bert.rst]:
```suggestion
This page shows how you can use models like BERT and GPT-2 in your contextual AI assitant
and includes practical tips on how to get the most out of these models.

```

File: docs/nlu/using-bert.rst [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+.. _using_bert:
+
+
+.. edit-link::
+
+Setup
+-----
+
+To demonstrate how to use ``BERT`` we will train two pipelines on Sara,
+the demo bot at Rasa. In doing this we will also be able to measure
```
Comment (amn41) [inline docs/nlu/using-bert.rst]:
```suggestion
the demo bot in the Rasa docs. In doing this we will also be able to measure
```

File: docs/nlu/using-bert.rst [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+In both cases we're training a :ref:`diet-classifier` for combined intent classification and entity recognition
+for 50 epochs but there are a few differences.
+
+In the light configuration we have :ref:`CountVectorsFeaturizer` which creates bag-of-word
+representations for each incoming message(at word and character levels). The heavy configuration replaces it with a
+``BERT`` model inside the pipeline. :ref:`HFTransformersNLP` is a utility component that does the heavy lifting work of loading the
+``BERT`` model in memory. Under the hood it leverages HuggingFace's `Transformers library <https://huggingface.co/transformers/>`_ to initialize the specified language model.
+Notice that we add two additional components :ref:`LanguageModelTokenizer` and :ref:`LanguageModelFeaturizer` which
+pick up the tokens and feature vectors respectively that are constructed by the utility component.
+
```
Comment (amn41) [inline docs/nlu/using-bert.rst]:
```suggestion
In both cases, the features extracted from the text (either a bag-of-words or BERT embeddings) are passed to the DIET model to train.
```

File: docs/nlu/using-bert.rst [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+.. note::
+
+    We strictly use these language models as featurizers, which means that their parameters are not fine-tuned during training of
+    downstream models in your NLU pipeline.
```
Comment (amn41) [inline docs/nlu/using-bert.rst]: explain why

File: docs/nlu/using-bert.rst (hunk 10 above)
Comment (koaning) [reply docs/nlu/using-bert.rst]: It'd be nice to have a proper link to huggingface somewhere tho.

**Comment by koaning** [comment]
@amn41 @dakshvar22 I think I've included all your comments. Is this now good to go?

File: docs/nlu/using-bert.rst [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+``BERT`` model inside the pipeline. :ref:`HFTransformersNLP` is a utility component that does the heavy lifting work of loading the
+``BERT`` model in memory. Under the hood it leverages HuggingFace's `Transformers library <https://huggingface.co/transformers/>`_ to initialize the specified language model.
+Notice that we add two additional components :ref:`LanguageModelTokenizer` and :ref:`LanguageModelFeaturizer` which
+pick up the tokens and feature vectors respectively that are constructed by the utility component.
+
+
+.. note::
+
+    We strictly use these language models as featurizers, which means that their parameters are not fine-tuned during training of
+    downstream models in your NLU pipeline. This is saves a lot of compute time and the machine learning models
```
Comment (dakshvar22) [inline docs/nlu/using-bert.rst:96]:
```suggestion
    downstream models in your NLU pipeline. This saves a lot of compute time and the machine learning models 
```

File: docs/nlu/using-bert.rst [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+Light       0.7818      0.7282      0.7448
+Heavy       0.8942      0.7642      0.8188
+========  =========== =========== ===========
+
+Observations 
+~~~~~~~~~~~~
+
+On all fronts we see that the heavy model with the ``BERT`` embeddings performs better.
+More specifically, the performance gain for intent classification is marginal but it deserves mentioning that the effect
+is more pronounced in the case of entity recognition.
```
Comment (amn41) [inline docs/nlu/using-bert.rst:172]: "it deserves mentioning that the effect is more pronounced" is not friendly English :) I'm sure there's a simpler way to say this.
